Allow GLideNUI to work with mupen64plus

Rosalie241 opened the pull request:
I want to be able to use upstream GLideN64 with GLideNUI on a custom mupen64plus (i.e m64p by Logan).

tim241 commented:
I think it's ready for review now

loganmc10 commented:
I'll do some testing on this, it would be nice to have this capability upstream.

One question, what is the purpose of the fullscreen resolution code for Unix/Linux? I don't think the fullscreen resolution, or refresh rate setting, are actually used anywhere for mupen64plus, they are only used in windows_DisplayWindow.cpp (zilmar spec). It kind of looks like dead code, unless I am missing something

tim241 commented:
> I'll do some testing on this, it would be nice to have this capability upstream.
> 
> One question, what is the purpose of the fullscreen resolution code for Unix/Linux? I don't think the fullscreen resolution, or refresh rate setting, are actually used anywhere for mupen64plus, they are only used in windows_DisplayWindow.cpp (zilmar spec). It kind of looks like dead code, unless I am missing something

hmm, yeah, it seems like 'dead code' at the moment, however, I implemented the fullscreen resolution code to make sure I didn't need to change the UI *at all*, anyways, wouldn't it be possible to implement support for it in `mupen64plus_DisplayWindow.cpp`? Maybe it'd be possible to hook it up with xlib for unix platforms?

loganmc10 commented:
@tim241 if you're going to attempt to implement the full screen resolution options in mupen64plus, you should use the builtin API function: ```CoreVideo_ListFullscreenModes``` to determine the fullscreen resolution options.

Documentation for the function is here:
https://github.com/mupen64plus/mupen64plus-core/blob/master/doc/emuwiki-api-doc/Mupen64Plus-v2.0-Core-Video-Extension.mediawiki

Basically you call that function and it provides the full screen resolution options, that way the platform specific code lives in the frontend, not the plugin. This is the way.

Rosalie241 commented:
Considering the mupen64plus PR has been merged, I've modified the commit's description and changed the patch a tiny bit to clean it up, I think it's ready for review now!

gonetz commented:
@loganmc10 could you please do the review?

loganmc10 commented:
I don't think I'll have time to test but I'll make some inline notes

loganmc10 commented on src/CMakeLists.txt:
```diff
@@ -152,13 +180,23 @@ if(MUPENPLUSAPI)
   )
 
+  if(MUPENPLUSAPI_GLIDENUI)
+      add_definitions(-DGLIDENUI)
```
I would rename this define to something like M64P_GLIDENUI to make it clear it's only related to mupen64plus+GLideNUI

Rosalie241 replied:
Fixed

loganmc10 commented on src/Graphics/OpenGLContext/mupen64plus/mupen64plus_DisplayWindow.cpp:
```diff
@@ -173,7 +179,32 @@ bool DisplayWindowMupen64plus::_resizeWindow()
 
 void DisplayWindowMupen64plus::_changeWindow()
 {
+#ifdef GLIDENUI
```
I think it would be nice to have a separate define for GLIDENUI_SUPPORTS_FULLSCREEN or something like that. Full screen resolution detection is pretty platform specific and might not work well on all frontends.

Rosalie241 replied:
I think it'd be saner if gliden64 would check if that function returns `M64ERR_UNSUPPORTED` and if it does, fall back to the original code

loganmc10 replied:
Yes, in ```DisplayWindowMupen64plus::_start()``` you could try to execute ```FunctionWrapper::CoreVideo_SetVideoModeWithRate```, if that returns ```M64ERR_UNSUPPORTED```, that tells you that the frontend doesn't support full screen resolution setting, and the rest of the code should execute based on that knowledge

Rosalie241 replied:
Implemented.
